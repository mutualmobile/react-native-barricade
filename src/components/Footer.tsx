import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale } from '../utils';

export type FooterProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
  titleStyle: TextStyle;
};

const Footer = (props: FooterProps): JSX.Element => {
  const { onPress, title, titleStyle } = props;
  const { themeColorStyle } = useThemedColor();

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      <TouchableOpacity onPress={onPress} style={styles.footer}>
        <Text numberOfLines={1} style={[styles.title, titleStyle]}>
          {title}
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 50,
    borderTopWidth: 1,
  },
  footer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: hScale(12),
  },
  title: {
    textAlign: 'center',
    fontSize: hScale(18),
  },
});

export { Footer };
