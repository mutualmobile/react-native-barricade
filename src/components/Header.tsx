import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';

export type ButtonProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
};

export type HeaderProps = {
  headerLeft?: ButtonProps;
  headerRight?: ButtonProps;
  title: string;
};

const Header = (props: HeaderProps): JSX.Element => {
  const { headerLeft, headerRight, title } = props;
  const { themeColorStyle } = useThemedColor();

  const renderButton = (buttonProps: ButtonProps, style: ViewStyle) => {
    return (
      <TouchableOpacity onPress={buttonProps.onPress} style={style}>
        <Text style={[styles.buttonText, themeColorStyle.primary]}>
          {buttonProps.title}
        </Text>
      </TouchableOpacity>
    );
  };

  const renderHeader = () => {
    return (
      <View style={styles.header}>
        <Text
          numberOfLines={1}
          style={[styles.title, themeColorStyle.textDark]}>
          {title}
        </Text>
        {headerLeft && renderButton(headerLeft, styles.headerLeft)}
        <View style={styles.flex} />
        {headerRight && renderButton(headerRight, styles.headerRight)}
      </View>
    );
  };

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      {renderHeader()}
    </View>
  );
};

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    borderBottomWidth: 1,
  },
  header: {
    marginHorizontal: hScale(12),
    height: vScale(50),
    alignItems: 'center',
    flexDirection: 'row',
  },
  title: {
    position: 'absolute',
    textAlign: 'center',
    fontSize: hScale(20),
    fontWeight: 'bold',
    marginHorizontal: hScale(60),
    left: 0,
    right: 0,
  },
  buttonText: {
    fontSize: hScale(18),
    margin: hScale(8),
    textAlign: 'center',
  },
  headerLeft: {
    marginRight: hScale(10),
  },
  headerRight: {
    marginLeft: hScale(10),
  },
});

export { Header };
