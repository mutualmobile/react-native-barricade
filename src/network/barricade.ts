//@ts-ignore
import * as RNFetch from 'react-native/Libraries/Network/fetch';

import { ObjectUtils, UrlUtils } from '../utils';
import {
  Method,
  PathEvaluationCallback,
  PathEvaluationType,
  Request,
  RequestConfig,
  RequestConfigForLib,
  ResponseData,
  ResponseHandler,
} from './barricade.types';
import { interceptor } from './interceptor';
import { createNativeXMLHttpRequest } from './xml-http-request';

export class Barricade {
  running = false;
  private readonly _originalRequestConfig: RequestConfigForLib[] = [];
  private _requestConfig: RequestConfigForLib[] = [];
  private _nativeXMLHttpRequest: typeof XMLHttpRequest;
  private _nativeFetch: RNFetch.fetch;
  private _nativeHeaders: RNFetch.Headers;
  private _nativeRequest: RNFetch.Request;
  private _nativeResponse: RNFetch.Response;

  constructor(requestConfig: RequestConfig[]) {
    this.initRequestConfig(requestConfig);
    this._originalRequestConfig = ObjectUtils.cloneDeep(this.requestConfig);
    this._nativeXMLHttpRequest = global.XMLHttpRequest;
    this._nativeFetch = global.fetch;
    this._nativeHeaders = global.Headers;
    this._nativeRequest = global.Request;
    this._nativeResponse = global.Response;
  }

  set requestConfig(value: RequestConfigForLib[]) {
    this._requestConfig = value;
  }

  get requestConfig() {
    return this._requestConfig;
  }

  handleRequest(request: Request) {
    const method = request._method.toUpperCase() as Method;
    const requestUrl = request._url;
    const parsedRequestUrl = UrlUtils.parseURL(requestUrl);
    request.params = parsedRequestUrl.params;

    const requestConfig = this._requestConfig.find(item => {
      if (
        item.method !== method ||
        !requestUrl.includes(item.pathEvaluation.path)
      ) {
        return;
      } else if (item.pathEvaluation?.type === PathEvaluationType.Includes) {
        return true;
      } else if (item.pathEvaluation?.type === PathEvaluationType.Suffix) {
        return requestUrl.endsWith(item.pathEvaluation.path);
      } else {
        return (item.pathEvaluation as PathEvaluationCallback).callback(
          request,
        );
      }
    });

    if (requestConfig) {
      try {
        const result = (
          requestConfig.responseHandler.find(item => !!item.isSelected) ??
          requestConfig.responseHandler[0]
        ).handler(request);
        if (
          result &&
          typeof (result as PromiseLike<ResponseData>)?.then === 'function'
        ) {
          (result as PromiseLike<ResponseData>).then(resolvedResult => {
            this.resolveRequest(request, resolvedResult, requestConfig.delay);
          });
        } else {
          this.resolveRequest(
            request,
            result as ResponseData,
            requestConfig.delay,
          );
        }
      } catch (error) {
        throw `Barricade intercepted ${requestUrl}(${method}) API and threw an error - ${
          (error as Error).message
        }.`;
      }
    } else {
      createNativeXMLHttpRequest(request, this._nativeXMLHttpRequest);
    }
  }

  resetRequestConfig() {
    this._requestConfig = ObjectUtils.cloneDeep(this._originalRequestConfig);
  }

  private resolveRequest(
    request: Request,
    responseData: ResponseData,
    delay = 400,
  ) {
    let response = responseData.response;
    if (request.responseType === 'blob') {
      // @ts-ignore
      response = (responseData.response as Blob)._data;
    }
    setTimeout(() => {
      request.setResponseData(
        responseData.status,
        responseData.headers,
        response,
      );
    }, delay);
  }

  start() {
    global.XMLHttpRequest = interceptor(this) as any;
    global.fetch = RNFetch.fetch;
    global.Headers = RNFetch.Headers;
    global.Request = RNFetch.Request;
    global.Response = RNFetch.Response;
    this.running = true;
  }

  shutdown() {
    global.XMLHttpRequest = this._nativeXMLHttpRequest;
    global.fetch = this._nativeFetch;
    global.Headers = this._nativeHeaders;
    global.Request = this._nativeRequest;
    global.Response = this._nativeResponse;
    this.running = false;
  }

  initRequestConfig(requests: RequestConfig[]) {
    const updatedRequestConfig = requests.map<RequestConfig>(request => {
      let selectedItem: ResponseHandler | undefined;
      for (let i = 0; i < request.responseHandler.length; i++) {
        if (selectedItem) {
          request.responseHandler[i].isSelected = false;
        } else if (request.responseHandler[i].isSelected) {
          selectedItem = request.responseHandler[i];
        } else if (i === request.responseHandler.length - 1) {
          request.responseHandler[i].isSelected = false;
          request.responseHandler[0].isSelected = true;
          selectedItem = request.responseHandler[0];
        } else {
          request.responseHandler[i].isSelected = false;
        }
      }
      const result = request as RequestConfigForLib;
      result.selectedResponseLabel = selectedItem!.label;

      return result;
    });

    this.requestConfig = updatedRequestConfig;
  }
}
