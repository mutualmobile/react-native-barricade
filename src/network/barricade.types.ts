import { HttpStatusCodeText } from './http-codes';
import { MockedXMLHttpRequest } from './mocked-xml-http-request';

export enum Method {
  Delete = 'DELETE',
  Get = 'GET',
  Head = 'HEAD',
  Options = 'OPTIONS',
  Patch = 'PATCH',
  Post = 'POST',
  Put = 'PUT',
}

export type RequestConfig = {
  path: string;
  method: Method;
  responseHandler: ResponseHandler;
  delay?: number;
};

export type ResponseData = {
  status: keyof typeof HttpStatusCodeText;
  headers: { [k: string]: string };
  response: string;
};

interface ExtraRequestData {
  _url: string;
  _method: string;
  params?: Record<string, string>;
}

export type MockedRequest = MockedXMLHttpRequest & ExtraRequestData;

export type ResponseHandler = {
  (request: MockedRequest): ResponseData | PromiseLike<ResponseData>;
};

export type RequestConfigForMethod = {
  [key in Method]: RequestConfig;
};
export type RequestReferences = Record<string, RequestConfigForMethod>;
