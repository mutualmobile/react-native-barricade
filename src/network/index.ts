export * from './barricade';
export * from './barricade.types';
export * from './http-codes';
