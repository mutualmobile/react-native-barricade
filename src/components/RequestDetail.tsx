import React from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import {
  Barricade,
  RequestConfig,
  RequestConfigForLib,
  ResponseHandlerForLib,
} from '../network';
import { useThemedColor } from '../theme';
import { Header } from './Header';

const CheckMark = '\u2713';
const ChevronLeft = '\u2039';

const RequestDetail = ({
  barricade,
  detailData,
  onBackPressed,
}: {
  barricade: Barricade | undefined;
  detailData?: RequestConfigForLib;
  onBackPressed: () => void;
}): JSX.Element => {
  const { themeColorStyle } = useThemedColor();

  const onDeailItemPressed = (item: ResponseHandlerForLib) => {
    if (!item.isSelected) {
      detailData?.responseHandler.map(responseHandler => {
        responseHandler.isSelected = responseHandler.label === item.label;
        return responseHandler;
      });
      const requestConfig = barricade?.requestConfig.map<RequestConfig>(
        config => {
          if (config.label === detailData?.label) {
            config = detailData;
          }
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { selectedResponseLabel, ...result } = config;
          return result;
        },
      );
      if (requestConfig) {
        barricade?.updateRequestConfig(requestConfig);
      }
      onBackPressed();
    }
  };

  const renderDetailItem = ({
    item,
  }: ListRenderItemInfo<ResponseHandlerForLib>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onDeailItemPressed(item)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        {item.isSelected && (
          <Text style={[styles.icon, themeColorStyle.primary]}>
            {CheckMark}
          </Text>
        )}
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{ title: ChevronLeft + ' Back', onPress: onBackPressed }}
        title={detailData?.label ?? ''}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={detailData?.responseHandler}
        renderItem={renderDetailItem}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 20,
    paddingVertical: 12,
    borderBottomWidth: 1,
  },
  label: {
    fontSize: 18,
    fontWeight: '400',
  },
  icon: {
    fontSize: 18,
  },
});

export { RequestDetail };
