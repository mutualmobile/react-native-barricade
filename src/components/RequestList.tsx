import React, { useState } from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { Strings, Unicode } from '../constants';
import { Barricade, RequestConfigForLib } from '../network';
import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';
import { Footer } from './Footer';
import { Header } from './Header';

type RequestListProps = {
  barricade: Barricade;
  onDonePressed?: (event: NativeSyntheticEvent<any>) => void;
  onListItemPressed: (index: number) => void;
};

const RequestList = (props: RequestListProps): JSX.Element => {
  const { barricade, onDonePressed, onListItemPressed } = props;
  const { themeColorStyle } = useThemedColor();
  const [refreshList, setRefreshList] = useState(0);

  const onResetPressed = () => {
    barricade?.resetRequestConfig();
    setRefreshList(Math.random());
  };

  const renderListItem = ({
    item,
    index,
  }: ListRenderItemInfo<RequestConfigForLib>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onListItemPressed(index)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        <Text style={[styles.value, themeColorStyle.textLight]}>
          {item.selectedResponseLabel}{' '}
          <Text style={[styles.icon, themeColorStyle.textLight]}>
            {Unicode.ChevronRight}
          </Text>
        </Text>
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{ title: Strings.Reset, onPress: onResetPressed }}
        title={Strings.Barricade}
        headerRight={{ title: Strings.Done, onPress: onDonePressed }}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={barricade?.requestConfig}
        renderItem={renderListItem}
        extraData={refreshList}
      />
      <Footer barricade={barricade} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: hScale(20),
    paddingVertical: vScale(12),
    borderBottomWidth: 1,
  },
  label: {
    fontSize: hScale(18),
    fontWeight: '400',
  },
  value: {
    fontSize: hScale(16),
  },
  icon: {
    fontSize: hScale(18),
  },
});

export { RequestList };
