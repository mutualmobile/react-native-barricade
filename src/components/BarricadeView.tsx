import React, { useState } from 'react';
import {
  Modal,
  NativeSyntheticEvent,
  SafeAreaView,
  StyleSheet,
} from 'react-native';

import { Barricade } from '../network';
import { ThemeProvider, ThemeType, useThemedColor } from '../theme';
import { RequestDetail } from './RequestDetail';
import { RequestList } from './RequestList';

enum ViewType {
  List,
  Detail,
}

export const BarricadeView = ({
  barricade,
  onRequestClose,
  theme = 'light',
  visible,
}: {
  barricade?: Barricade;
  onRequestClose?: (event: NativeSyntheticEvent<any>) => void;
  theme?: ThemeType;
  visible: boolean;
}) => {
  const { themeColorStyle } = useThemedColor();
  const [viewType, setViewType] = useState<ViewType>(ViewType.List);
  const [selectedListItemIndex, setSelectedListItemIndex] = useState<number>();

  const _onBackPressed = () => {
    setSelectedListItemIndex(undefined);
    setViewType(ViewType.List);
  };

  const _onListItemPressed = (index: number) => {
    setSelectedListItemIndex(index);
    setViewType(ViewType.Detail);
  };

  const renderContent = () => {
    if (viewType === ViewType.List) {
      return (
        <RequestList
          barricade={barricade}
          onListItemPressed={_onListItemPressed}
          onDonePressed={onRequestClose}
        />
      );
    } else {
      return (
        <RequestDetail
          barricade={barricade}
          selectedListItemIndex={selectedListItemIndex}
          onBackPressed={_onBackPressed}
        />
      );
    }
  };

  return (
    <ThemeProvider value={theme}>
      <Modal onRequestClose={onRequestClose} visible={visible}>
        <SafeAreaView
          style={[
            styles.container,
            themeColorStyle.background,
            themeColorStyle.border,
          ]}>
          {renderContent()}
        </SafeAreaView>
      </Modal>
    </ThemeProvider>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
