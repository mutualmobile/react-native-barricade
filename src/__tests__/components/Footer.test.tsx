import React from 'react';
import { TouchableOpacity } from 'react-native';
import renderer, { act } from 'react-test-renderer';

import { Footer } from '../../components/Footer';
import { Barricade } from '../../network/barricade';
import {
  firstApiConfig,
  secondApiConfig,
  thirdApiConfig,
} from '../data/barricade-test-data';

let barricade: Barricade;
beforeEach(() => {
  barricade = new Barricade([firstApiConfig, secondApiConfig, thirdApiConfig]);
  barricade.start();
});

afterEach(() => {
  jest.clearAllMocks();
  barricade.shutdown();
});

describe('given that Barricade is enabled,', () => {
  test('should match Footer snapshot', () => {
    const tree = renderer.create(<Footer barricade={barricade} />);
    expect(tree).toMatchSnapshot();
    expect.assertions(1);

    barricade.shutdown();
  });

  test('when "Disable barricade" button is pressed, should disable barricade', async () => {
    barricade.shutdown = jest.fn();
    const tree = renderer.create(<Footer barricade={barricade} />);
    const instance = tree.root;
    await act(() => instance.findByType(TouchableOpacity).props.onPress());

    expect(barricade.shutdown).toHaveBeenCalled();
    expect.assertions(1);
  });
});

describe('given that Barricade is disabled,', () => {
  test('given that Barricade is disabled, should match Footer snapshot', () => {
    barricade.shutdown();

    const tree = renderer.create(<Footer barricade={barricade} />);
    expect(tree).toMatchSnapshot();

    expect.assertions(1);
  });

  test('when "Enable barricade" button is pressed, should enable barricade', async () => {
    barricade.shutdown();
    barricade.start = jest.fn();
    const tree = renderer.create(<Footer barricade={barricade} />);
    const instance = tree.root;
    await act(() => instance.findByType(TouchableOpacity).props.onPress());

    expect(barricade.start).toHaveBeenCalledTimes(1);
    expect.assertions(1);
  });
});
