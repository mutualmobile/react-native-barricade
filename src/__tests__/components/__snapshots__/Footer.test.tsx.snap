// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match Footer snapshot 1`] = `
<View
  style={
    [
      {
        "borderTopWidth": 1,
        "height": 50,
      },
      {
        "backgroundColor": "#EBEBEB",
      },
      {
        "borderColor": "#DEDEDE",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginHorizontal": 24,
        "opacity": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        [
          {
            "fontSize": 36,
            "textAlign": "center",
          },
          {
            "color": "#EE0003",
          },
        ]
      }
    >
      Disable Barricade
    </Text>
  </View>
</View>
`;
